datahub-actions/src/datahub_actions/plugin/source/acryl/datahub_cloud_event_source.py

```diff
@@ -0,0 +1,202 @@
+import json
+import logging
+import time
+from dataclasses import dataclass
+from typing import Iterable, List, Optional
+
+from datahub.configuration import ConfigModel
+from datahub.emitter.serialization_helper import post_json_transform
+
+# DataHub imports.
+from datahub.metadata.schema_classes import GenericPayloadClass
+
+from datahub_actions.event.event_envelope import EventEnvelope
+from datahub_actions.event.event_registry import (
+    ENTITY_CHANGE_EVENT_V1_TYPE,
+    EntityChangeEvent,
+)
+
+# May or may not need these.
+from datahub_actions.pipeline.pipeline_context import PipelineContext
+from datahub_actions.plugin.source.acryl.datahub_cloud_events_ack_manager import (
+    AckManager,
+)
+from datahub_actions.plugin.source.acryl.datahub_cloud_events_consumer import (
+    DataHubEventsConsumer,
+    ExternalEvent,
+)
+from datahub_actions.source.event_source import EventSource
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+ENTITY_CHANGE_EVENT_NAME = "entityChangeEvent"
+
+
+# Converts a DataHub Events Message to an EntityChangeEvent.
+def build_entity_change_event(payload: GenericPayloadClass) -> EntityChangeEvent:
+    return EntityChangeEvent.from_json(payload.get("value"))
+
+
+class DataHubEventsSourceConfig(ConfigModel):
+    topic: str = "PlatformEvent_v1"
+    consumer_id: Optional[str]  # Used to store offset for the consumer.
+    lookback_days: Optional[int] = None
+    reset_offsets: Optional[bool] = False
+
+    # Time and Exit Conditions.
+    kill_after_idle_timeout: bool = False
+    idle_timeout_duration_seconds: int = 30
+    event_processing_time_max_duration_seconds: int = 30
+
+
+# This is the custom DataHub-based Event Source.
+@dataclass
+class DataHubEventSource(EventSource):
+    running = False
+    source_config: DataHubEventsSourceConfig
+    ctx: PipelineContext
+
+    @staticmethod
+    def _get_pipeline_urn(pipeline_name: str) -> str:
+        if pipeline_name.startswith("urn:li:dataHubAction:"):
+            return pipeline_name
+        else:
+            return f"urn:li:dataHubAction:{pipeline_name}"
```

Comment:
prefer urn_builder here

Reply:
This is a pretty rarely constructured urn from python, so we don't have a utility for it: https://github.com/datahub-project/datahub/blob/8ac35fa9ee6229ab7b8c0f4b2a0e669113574cd1/metadata-ingestion/src/datahub/emitter/mce_builder.py#L125 (unless you're referring to another set of utils)